File: C/Greedy-Algorithm/fractional_knapsack.c
#include <stdio.h>

void main()
{
    // capacity = knapsack capacity
    // no_items:- number of items that can be enter
    // cur_weight:- current weight
    int capacity, no_items, cur_weight, item;
    int used[10]; // to keep track of which items have been added to the knapsack
    float total_profit;
    int i;
    int weight[10];
    int value[10];

    printf("Enter the capacity of knapsack:\n");
    scanf("%d", &capacity);

    printf("Enter the number of items:\n");
    scanf("%d", &no_items);

    printf("Enter the weight and value of %d item:\n", no_items);
    for (i = 0; i < no_items; i++)
    {
        // enter weight and value of each object
        printf("Weight[%d]:\t", i);
        scanf("%d", &weight[i]);
        printf("Value[%d]:\t", i);
        scanf("%d", &value[i]);
    }

    for (i = 0; i < no_items; ++i)
        used[i] = 0;

    cur_weight = capacity; // to keep track of the current weight of the knapsack
    while (cur_weight > 0)
    {
        item = -1;
        for (i = 0; i < no_items; ++i)
            if ((used[i] == 0) &&
                ((item == -1) || ((float)value[i] / weight[i] > (float)value[item] / weight[item])))
                item = i;

        used[item] = 1;
        cur_weight -= weight[item];
        total_profit += value[item]; // calculates the total profit by adding the value of all the items added to the knapsack
        if (cur_weight >= 0)
            printf("Added object %d (%d Rs., %dKg) completely in the bag. Space left: %d.\n", item + 1, value[item], weight[item], cur_weight);
        else
        {
            int item_percent = (int)((1 + (float)cur_weight / weight[item]) * 100);
            printf("Added %d%% (%d Rs., %dKg) of object %d in the bag.\n", item_percent, value[item], weight[item], item + 1);
            total_profit -= value[item];
            total_profit += (1 + (float)cur_weight / weight[item]) * value[item];
        }
        // If the item can be added completely to the knapsack, it is added in its entirety, and the space remaining in the knapsack is updated. If the item cannot be added completely to the knapsack, only a fraction of the item is added, and the space remaining in the knapsack is updated accordingly.
    }

    printf("Filled the bag with objects worth %.2f Rs.\n", total_profit);
}

File: Java/Greedy-Algorithm/FractionalKnapsack.java
/* The Fractional Knapsack problem is a classic optimization problem often solved using a greedy algorithm. 
It involves selecting items from a set, each with a given weight and value, to maximize 
the total value while respecting a weight constraint.
Unlike the 0/1 knapsack problem, items can be divided (fractional) to maximize the value. 
This means that a fraction of an item can be included in the knapsack, rather than just selecting or excluding the entire item.*/

import java.util.Scanner;

class FractionalKnapsack {
    public static void main(String[] args) {
        float sum = 0;
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of items: ");
        int n = sc.nextInt();

        /* Creating an array to store weights and profits.
        Each column represents an item, and the first row stores the weights, while the second row stores the profits.*/
        int array[][] = new int[2][n];

        int i, j = 0;
        System.out.println("Enter the weight of each item:");
        for (i = 0; i < n; i++) {
            array[0][i] = sc.nextInt();
        }
        System.out.println("Enter the profit of each item:");
        for (i = 0; i < n; i++) {
            array[1][i] = sc.nextInt();
        }
        System.out.print("Enter maximum volume of knapsack: ");
        int totalWeight = sc.nextInt();

        int m = totalWeight;
        while (m >= 0) {

            /* Finding the item with the maximum profit-to-weight ratio */
            float max = 0;
            for (i = 0; i < n; i++) {
                if (((float) array[1][i]) / ((float) array[0][i]) > max) {
                    max = ((float) array[1][i]) / ((float) array[0][i]);
                    j = i;
                }
            }

            /* If the weight of the item (array[0][j]) is greater than the remaining capacity of the knapsack (m), 
            it means the item cannot be fully added. In that case, the program outputs a message indicating 
            the quantity of the item added and calculates the profit contribution (m * max) of the partially added item. 
            Then, it sets m to -1 to exit the while loop. */
            if (array[0][j] > m) {
                System.out.println("Quantity of item number " + (j + 1) + " added is " + m);
                sum += m * max;
                m = -1;
            }
            
            /* If the item can be fully added to the knapsack, the program outputs a message indicating 
            the quantity of the item added. It updates m by subtracting the weight of the added item, 
            subtracts the profit of the added item from the sum, and sets the profit of the added item to 0 to mark it as already added. */
            else {
                System.out.println("Quantity of item number " + (j + 1) + " added is " + array[0][j]);
                m -= array[0][j];
                sum += (float) array[1][j];
                array[1][j] = 0;
            }
        }

        /* Displays the total profit calculated */
        System.out.println("Total Profit " + sum);

        sc.close();
    }
}